MvvmCross-Forms/MvvmCross.Forms/Views/MvxFormsPagePresenter.cs

```diff
@@ -150,34 +151,61 @@ public override void RegisterAttributeTypes()
                 });
         }
 
-        public override void ChangePresentation(MvxPresentationHint hint)
+        public async override void ChangePresentation(MvxPresentationHint hint)
         {
             var navigation = GetPageOfType<NavigationPage>().Navigation;
             if (hint is MvxPopToRootPresentationHint popToRootHint)
             {
-                navigation.PopToRootAsync(popToRootHint.Animated);
+                // Make sure all modals are closed
+                CloseAllModals(popToRootHint.Animated);
+
+                // Double check we have a navigation page, otherwise
+                // we can just return as we must be already at the root page
+                if (navigation == null)
+                    return;
+
+                // Close all pages back to the root
+                await navigation.PopToRootAsync(popToRootHint.Animated);
                 return;
             }
             if (hint is MvxPopPresentationHint popHint)
             {
-                foreach (var page in navigation.NavigationStack)
+                // Need to check the modal stack first
+                while (navigation.ModalStack.Any())
                 {
-                    page.Navigation.PopAsync(popHint.Animated);
-                    if (page is IMvxPage mvxPage && mvxPage.ViewModel.GetType() == popHint.ViewModelToPopTo)
+                    var modalPage = navigation.ModalStack.Last();
+                    if (modalPage.PageMatchesViewModel(popHint.ViewModelToPopTo))
                         return;
+
+                    var modalNavPage = GetPageOfType<NavigationPage>(modalPage);
+                    if (modalNavPage != null)
+                    {
+                        var matched = await PopToMatchingViewModel(modalNavPage.Navigation, popHint.ViewModelToPopTo, popHint.Animated);
+                        if (matched)
+                            return;
+                    }
+
+                    await navigation.PopModalAsync();
                 }
+
+                await PopToMatchingViewModel(navigation, popHint.ViewModelToPopTo, popHint.Animated);
                 return;
             }
             if (hint is MvxRemovePresentationHint removeHint)
             {
+                if (navigation.ModalStack.Any())
```

Comment:
The page could be part of the navigationpage inside a modal? Or you might want to remove a page while a modal is showing.

Reply:
For some reason removing a page whilst modal is showing is broken on iOS. Did it ever work?